app/controllers/samples_controller.rb

```diff
@@ -56,47 +54,47 @@ def show
 
   def release
     @sample = Sample.find(params[:id])
-    redirect_if_not_owner_or_admin_otherwise do
-      if @sample.released?
-        flash[:notice] = "Sample '#{@sample.name}' already publically released"
-      else
-        @sample.release
-        flash[:notice] = "Sample '#{@sample.name}' publically released"
-      end
-      redirect_to sample_path(@sample)
+    authorize! :release, @sample
+
+    if @sample.released?
+      flash[:notice] = "Sample '#{@sample.name}' already publically released"
+    else
+      @sample.release
+      flash[:notice] = "Sample '#{@sample.name}' publically released"
     end
+    redirect_to sample_path(@sample)
   end
 
   def edit
     @sample = Sample.find(params[:id])
-    redirect_if_not_owner_or_admin_otherwise do
-      if @sample.released? && !current_user.administrator?
-        flash[:error] = 'Cannot edit publically released sample'
-        redirect_to sample_path(@sample)
-        return
-      end
+    authorize! :update, @sample
 
-      respond_to do |format|
-        format.html
-        format.xml  { render xml: @samples.to_xml }
-        format.json { render json: @samples.to_json }
-      end
+    if @sample.released? && cannot?(:update_released, @sample)
```

Comment:
I had a few situations like this, where an ability to do something was dependant on both the state of the object, as well as the user. While cancancan does let you put these conditions in the object itself, I avoided that in most cases. I decided that in this case the ability to release a sample that was already released was constrained by reality, not the person doing it, so the logic didn't belong in the permissions system.